src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp

```diff
@@ -0,0 +1,191 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+// IWYU pragma: no_include <array>
+
+#include "Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp"
+#include "Characteristics.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp" // IWYU pragma: keep
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp" // IWYU pragma: keep
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Tags::CharSpeed
+// IWYU pragma: no_forward_declare Tags::Psi
+// IWYU pragma: no_forward_declare Tags::Pi
+// IWYU pragma: no_forward_declare Tags::Phi
+// IWYU pragma: no_forward_declare Tags::UPsi
+// IWYU pragma: no_forward_declare Tags::UZero
+// IWYU pragma: no_forward_declare Tags::UMinus
+// IWYU pragma: no_forward_declare Tags::UPlus
+
+namespace GeneralizedHarmonic {
+
+template <size_t Dim, typename Frame>
+typename Tags::CharacteristicSpeeds<Dim, Frame>::type
+CharacteristicSpeedsCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_1, const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, Dim, Frame>& shift,
+    const tnsr::i<DataVector, Dim, Frame>& normal) noexcept {
+  auto char_speeds =
+      make_with_value<typename Tags::CharacteristicSpeeds<Dim, Frame>::type>(
+          get(lapse), 0.);
+  const auto shift_dot_normal = get(dot_product(shift, normal));
+  get(get<::Tags::CharSpeed<Tags::UPsi<Dim, Frame>>>(char_speeds)) =
+      -(1. + get(gamma_1)) * shift_dot_normal;
+  get(get<::Tags::CharSpeed<Tags::UZero<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal;
+  get(get<::Tags::CharSpeed<Tags::UPlus<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal + get(lapse);
+  get(get<::Tags::CharSpeed<Tags::UMinus<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal - get(lapse);
+
+  return char_speeds;
+}
+
+
+template <size_t Dim, typename Frame>
+typename Tags::EvolvedFieldsFromCharacteristicFields<Dim, Frame>::type
+EvolvedFieldsFromCharacteristicFieldsCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim, Frame>& u_psi,
+    const tnsr::iaa<DataVector, Dim, Frame>& u_zero,
+    const tnsr::aa<DataVector, Dim, Frame>& u_plus,
+    const tnsr::aa<DataVector, Dim, Frame>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame>& unit_normal_one_form) noexcept {
+  auto evolved_fields =
+      make_with_value<typename Tags::EvolvedFieldsFromCharacteristicFields<Dim,
+          Frame>::type>(get(gamma_2), 0.);
+
+  // Invert Eq.(32) of Lindblom+ (2005) for Psi
+  get<Tags::Psi<Dim, Frame>>(evolved_fields) = u_psi;
+
+  for (size_t mu = 0; mu < Dim + 1; ++mu) {
```

Comment:
We usually use a, b, c, ... for spacetime indices in spectre. Since you use these in the help text, please make consistent here and throughout the PR. Not a big deal.

Reply:
fixed